cobbler/tftpgen.py

```diff
@@ -1783,3 +1965,109 @@ def _write_bootcfg_file(self, what: str, name: str, filename: str) -> str:
                 raise OSError(f"Error creating symlink {link_file}") from os_error
 
         return buffer
+
+    def _read_chunk(
+        self, path: Union[str, pathlib.Path], offset: int, size: int
+    ) -> Tuple[bytes, int]:
+        with open(path, "rb") as fd:
+            fd.seek(offset)
+            data = fd.read(size)
+            file_size = fd.seek(0, os.SEEK_END)
+            return data, file_size
+
+    def _get_static_tftp_file(
+        self, path: pathlib.Path, offset: int, size: int
+    ) -> Optional[Tuple[bytes, int]]:
+        relative_path = str(path).strip("/")
+        try:
+            return self._read_chunk(
+                pathlib.Path(self.settings.bootloaders_dir, relative_path), offset, size
+            )
+        except FileNotFoundError:
+            try:
+                return self._read_chunk(
+                    pathlib.Path(self.settings.grubconfig_dir, relative_path),
+                    offset,
+                    size,
+                )
+            except FileNotFoundError:
+                return None
+
+    def _get_distro_tftp_file(
+        self, path: pathlib.Path, offset: int, size: int
+    ) -> Optional[Tuple[bytes, int]]:
+        if path.parts[1] != "images":
+            distro_name = path.parts[3]
+        else:
+            distro_name = path.parts[2]
+        distro = self.api.find_distro(distro_name, return_list=False)
+        if isinstance(distro, list):
+            raise ValueError("Expected a single distro, found a list")
+        if distro is not None:
+            kernel_path = pathlib.Path(distro.kernel)
+            if path.name == kernel_path.name:
+                return self._read_chunk(kernel_path, offset, size)
+            initrd_path = pathlib.Path(distro.initrd)
+            if path.name == initrd_path.name:
+                return self._read_chunk(initrd_path, offset, size)
+        return None
+
+    def _generate_tftp_config_file(
+        self, path: pathlib.Path, offset: int, size: int
+    ) -> Optional[Tuple[bytes, int]]:
+        metadata = self.get_menu_items()
+        # TODO: This iterates through all systems for each request. Can this be optimized?
+        for system in self.api.systems():
+            if not system.is_management_supported():
+                continue
+            contents = self.generate_system_file(system, path, metadata)
+            if contents is not None:
+                break
+        if contents is None:
+            contents = self.generate_pxe_menu(path, metadata)
+        if contents is not None:
+            enc = contents.encode("UTF-8")
+            return enc[offset:offset+size], len(enc)
```

Comment:
The annotation from Codacy is indeed correct you can run into an unbound error in case there are no systems added in Cobbler (which is a common case).